Add complete status

mayura-andrew opened the pull request:
## Purpose
<!--- Describe the problems, issues, or needs driving this feature/fix and include links to related issues -->
The purpose of this PR is to fix #121 
## Goals
<!---  Describe the solutions that this feature/fix will introduce to resolve the problems described above -->
 Once the program concludes, the admin should have the ability to update the status of each mentee's application to 'COMPLETE'.

## Approach
<!--- Describe how you are implementing the solutions. Include a link to a Markdown file or Google doc if the feature write-up is too long to paste here. -->
Creating new separate UI component and use it.

## Screenshots
<!--- Attach screenshots or demo-gif of any UI changes -->
![image](https://github.com/sef-global/scholarx-frontend/assets/48531182/86e8baf7-c894-4d3f-bd60-22be3051c1b4)
![image](https://github.com/sef-global/scholarx-frontend/assets/48531182/0bd75eba-dcce-4cd3-85d6-2e8fcaa0a3f2)
![image](https://github.com/sef-global/scholarx-frontend/assets/48531182/a7fb9415-2a6f-4fbd-99bb-82f287990490)
![image](https://github.com/sef-global/scholarx-frontend/assets/48531182/27c55e22-ca2c-42cf-99e0-c85d01f6a0ce)

##  Checklist
- [X] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [X] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

## Related PRs
<!--- List any other related PRs --> 

## Test environment
<!--- List all JDK versions, operating systems, databases, and browser/versions on which this feature/fix was tested --> 

 - OS: Linux Mint 21.3
 - Browser Firefox
 - Version 128.0


## Learning
<!--- Describe the research phase and any blog posts, patterns, libraries, or add-ons you used to solve the problem. -->

mayura-andrew committed:
changed

mayura-andrew committed:
Implement Send Email Ui

mayura-andrew committed:
Change "Status" to "Sent Status"in EmailHistory

mayura-andrew committed:
Update src/pages/Dashboard/scenes/Emails/Emails.tsx

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

mayura-andrew committed:
Update src/components/Dashboard/scenes/Emails/CheckButton.tsx

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

mayura-andrew committed:
Update src/components/Dashboard/scenes/Emails/EmailTemplate.tsx

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

mayura-andrew committed:
Implement Email Feature

mayura-andrew committed:
Implement Google OAuth butoon

mayura-andrew committed:
Refactor

mayura-andrew committed:
Remove Array.isArray check in EmailHistory.tsx

mayura-andrew committed:
Update example.env

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

mayura-andrew committed:
Removed automatic email fetching

mayura-andrew committed:
Implement useFetchEmails hook

mayura-andrew committed:
Ui fixes for email history and emails page

mayura-andrew committed:
Removed refetch in useFetchEmails

mayura-andrew committed:
Merge branch 'sef-global:main' into main

mayura-andrew committed:
Improved UI for Email History and Emails page

mayura-andrew committed:
Fixed Emails.tsx, EmailHistory.tsx

mayura-andrew committed:
Truncate Email History Table

mayura-andrew committed:
Merge branch 'main' into main

mayura-andrew committed:
Draft implement useSendEmail hook and send email functionality

mayura-andrew committed:
Implement useSendEmail hook for email operations

mayura-andrew committed:
Implement useSendEmail hook for email operations

mayura-andrew committed:
remove extra typoes

mayura-andrew committed:
Fixed mail history table UI issue

mayura-andrew committed:
Remove Extra variables from Emails.tsx file.

mayura-andrew committed:
fixed sendEmail mutation

mayura-andrew committed:
remove console.log() from useEmails.ts

mayura-andrew committed:
Change EmailHistory table sent, open status

mayura-andrew committed:
EmailHistory emails status

mayura-andrew committed:
Improved UI for Email History

mayura-andrew committed:
Merge branch 'main' into main

mayura-andrew committed:
Fixed conflicts

mayura-andrew committed:
Fixed Conflicts

anjula-sack committed:
Merge branch 'main' into main

mayura-andrew committed:
Removed esbuild package

mayura-andrew committed:
Merge branch 'main' of https://github.com/mayura-andrew/scholarx-frontend

mayura-andrew committed:
Improve UI

mayura-andrew committed:
Merge branch 'main' into main

mayura-andrew committed:
Fix Conflicts

mayura-andrew committed:
Merge branch 'sef-global:main' into main

mayura-andrew committed:
Compressed .png files in public folder

mayura-andrew committed:
Change Apply button text to Find a Mentor

mayura-andrew committed:
Merge branch 'sef-global:main' into bulk-email-send-feature

mayura-andrew committed:
Implement bulk email fetch from api

mayura-andrew committed:
Removed uneccessary imports

mayura-andrew committed:
Refactor Emails component to use ApplicationStatus enum

mayura-andrew committed:
Merge branch 'sef-global:main' into bulk-email-send-feature

mayura-andrew committed:
chore: Add password reset functionality and pages

mayura-andrew committed:
Improved response message

mayura-andrew committed:
Update types.ts and useForgotPassword.ts

mayura-andrew committed:
Add react-hook-form for form validation in ForgotPasswordModal

mayura-andrew committed:
Update password reset modal to display error and success messages

mayura-andrew committed:
chore: Update password reset modal to display error and success messages

mayura-andrew committed:
Add password show and hide

mayura-andrew committed:
Remove max-width from MenteeApplicationPage and MentorApplicationPage

mayura-andrew committed:
Refactor approve/reject button rendering in MenteeApplication, MenteeProfile, and MentorApplication components

mayura-andrew committed:
Refactor approve/reject button rendering in MentorApplications component

mayura-andrew committed:
Merge branch 'sef-global:main' into bulk-email-send-feature

mayura-andrew committed:
Add mentorship complete state

mayura-andrew committed:
Merge branch 'sef-global:main' into Add-complete-status

mayura-andrew committed:
Merge branch 'sef-global:main' into Add-complete-status

mayura-andrew committed:
feat: Update complete button to handle state change asynchronously

mayura-andrew committed:
feat: Update mentor availability message

Display a specific message when a mentor is not accepting new applications at the moment. This improves the user experience by providing clear information to potential mentees. The message is displayed in a yellow box with a rounded border.

Fixes #123

anjula-sack commented on src/enums.ts:
```diff
@@ -7,6 +7,7 @@ export enum ApplicationStatus {
   PENDING = 'pending',
   REJECTED = 'rejected',
   APPROVED = 'approved',
+  COMPLATED = 'completed',
```
```suggestion
  COMPLETED = 'completed',
```

mayura-andrew committed:
Update src/enums.ts

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,50 @@
+}) => {
+  const [isModalOpen, setIsModalOpen] = useState(false);
+  const [modalMessage, setModalMessage] = useState('');
+  const [onConfirmAction, setOnConfirmAction] = useState<() => void>(() => {});
```
```suggestion
```

anjula-sack replied:
Remove this as well

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,50 @@
+
+  const handleComplate = () => {
+    setModalMessage('Are you sure you want to complete?');
+    setOnConfirmAction(() => complete);
```
```suggestion
    complete()
```

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,50 @@
+        isOpen={isModalOpen}
+        onClose={handleCloseModal}
+        onConfirm={handleConfirm}
+        message={modalMessage}
```
```suggestion
        message="Are you sure you want to complete?"
```

mayura-andrew committed:
feat: Add complete mentorship button

This commit adds a new button to the mentorship application page that allows mentors to mark a mentorship as complete. The button is only displayed when the mentorship is in the approved state. Clicking the button opens a confirmation modal asking for confirmation before marking the mentorship as complete.

Fixes #123

anjula-sack commented on src/components/ApproveRejectButtons/index.tsx:
```diff
@@ -3,14 +3,18 @@ import ConfirmationModal from '../ConfirmationModal';
+  approve?: () => void;
+  reject?: () => void;
+  complete?: () => void;
+  canComplete: boolean;
```
We don't need this.

we can check `complete` is not undefined here.

```suggestion
```

anjula-sack commented on src/components/ApproveRejectButtons/index.tsx:
```diff
@@ -28,6 +32,12 @@ const ApproveRejectButtons: React.FC<ApproveRejectButtonsProps> = ({
 
+  const handleComplete = () => {
+    setModalMessage('Are you sure you want to complete?');
+    setOnConfirmAction(() => complete);
```
```suggestion
    setOnConfirmAction(complete);
```

mayura-andrew committed:
chore: Remove unused prop 'canComplete' from ApproveRejectButtons component

mayura-andrew committed:
chore: Remove unused prop 'canComplete' from MentorApplication component

mayura-andrew committed:
chore: Remove unused prop 'canComplete' from MenteeProfile component

mayura-andrew committed:
feat: Add complete mentorship button

anjula-sack commented on src/components/ApproveRejectButtons/index.tsx:
```diff
@@ -39,18 +39,21 @@ const ApproveRejectButtons: React.FC<ApproveRejectButtonsProps> = ({
-      >
-        {isLoading ? 'Loading...' : 'Reject'}
-      </button>
+      {' '}
```
```suggestion
```

mayura-andrew committed:
chore: Remove unnecessary whitespace in ApproveRejectButtons component

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,50 @@
+  const handleConfirm = () => {
+    onConfirmAction();
+    setIsModalOpen(false);
+  };
```
```suggestion
  const handleConfirm = () => {
    complete();
    setIsModalOpen(false);
  };
```

mayura-andrew committed:
Merge branch 'sef-global:main' into Add-complete-status

mayura-andrew committed:
Update src/components/CompleteButton/index.tsx

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

mayura-andrew committed:
Update src/components/CompleteButton/index.tsx

Co-authored-by: Anjula Shanaka <anjulashanaka@gmail.com>

mayura-andrew committed:
chore: Refactor CompleteButton component to improve code readability and maintainability

anjula-sack commented on package-lock.json:
```diff
@@ -14,6 +14,7 @@
         "@tanstack/react-query-devtools": "^5.37.1",
         "axios": "^1.4.0",
         "dotenv": "^16.0.3",
+        "flowbite-react": "^0.10.1",
```
Can you remove this?

mayura-andrew committed:
Refactor CompleteButton component to use complete function directly

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,46 @@
+
+interface CompleteButtonProps {
+  isLoading: boolean;
+  complete: () => void;
```
```suggestion
  complete: Promise<void>;
```

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,46 @@
+  };
+
+  const handleConfirm = () => {
+    complete();
```
```suggestion
    await complete();
```

mayura-andrew committed:
chore: Update CompleteButton component to use async/await for complete function

anjula-sack reviewed (approved):
Thank you!

anjula-sack commented on src/components/MenteeApplication/MenteeApplication.component.tsx:
```diff
@@ -89,6 +90,14 @@ const MenteeApplication: React.FC = () => {
             )}
+            {mentee?.state === ApplicationStatus.APPROVED && (
+              <CompleteButton
+                isLoading={isPending}
```
```suggestion
                isLoading={isLoading}
```

anjula-sack commented on src/components/CompleteButton/index.tsx:
```diff
@@ -0,0 +1,46 @@
+        onClose={handleCloseModal}
+        onConfirm={handleConfirm}
+        message="Are you sure you want to complete?"
+      />
```
```suggestion
       Complete Mentorship
      </button>
      <ConfirmationModal
        isOpen={isModalOpen}
        isLoading={isLoading}
        onClose={handleCloseModal}
        onConfirm={handleConfirm}
        message="Are you sure you want to complete?"
      />
```

anjula-sack commented on src/utils.ts:
```diff
@@ -6,6 +6,8 @@ export const getStateColor = (state: string | undefined) => {
     case 'rejected':
       return 'bg-red-100 text-red-700';
+    case 'completed':
+      return 'bg-yellow-100 text-yellow-700';
```
```suggestion
      return 'bg-green-100 text-green-700';
```

mayura-andrew committed:
chore: Add loading state to ConfirmationModal's confirm button

mayura-andrew committed:
Update CompleteButton component to use async/await for complete function

mayura-andrew committed:
chore: Remove unused 'isPending' variable in MenteeApplication component

mayura-andrew committed:
chore: Remove unused 'isPending' variable in MenteeApplication component

mayura-andrew committed:
Refactor MenteeProfile and MentorApplication components to use ActionButtons for state updates

mayura-andrew committed:
Refactor MenteeProfile and MentorApplication components to use ActionButtons for state updates

mayura-andrew committed:
Refactor ActionButtons component to improve code reusability and readability

anjula-sack commented on src/components/ConfirmationModal/index.tsx:
```diff
@@ -16,6 +16,13 @@ const ConfirmationModal: React.FC<ConfirmationModalProps> = ({
   message,
 }) => {
   if (!isOpen) return null;
+  const [buttonText, setButtonText] = useState('Confirm');
```
```suggestion
  const [isLoading, setIsLoading] = useState(false);
```

mayura-andrew committed:
Refactor ConfirmationModal component to use loading state for the confirm button

anjula-sack reviewed (approved):
Thank you! @mayura-andrew
